nameko/testing/services.py

```diff
@@ -48,93 +48,157 @@ def entrypoint_hook(container, method_name, context_data=None):
 
-    Usage::
+        worker_ctx (WorkerContext): WorkerContext of the entrypoint call.
 
-        container = ServiceContainer(ExampleService, config)
-        with entrypoint_waiter(container, 'example_handler'):
-            ...  # e.g. rpc call that will result in handler being called
-    """
+        result (object): The result, if any, that the entrypoint returned.
+
+        exc_info (tuple): Tuple as returned by `sys.exc_info` if the
+            entrypoint raised an exception, otherwise `None`.
 
-    waiter = EntrypointWaiter(entrypoint)
-    if not get_extension(container, Entrypoint, method_name=entrypoint):
+
+    The `timeout` argument specifies the maximum number of seconds the
+    `entrypoint_waiter` should wait before exiting. It can be disabled by
+    passing `None`.
+
+    **Usage**
+
+    ::
+        class Service(object):
+            name = "service"
+
+            @event_handler('srcservice', 'eventtype')
+            def handle_event(self, msg):
+                return msg
+
+        container = ServiceContainer(Service, config)
+        container.start()
+
+        # basic
+        with entrypoint_waiter(container, 'handle_event'):
+            ...  # action that dispatches event
+
+        # giving access to the result
+        with entrypoint_waiter(container, 'handle_event') as result:
+            ...  # action that dispatches event
+        res = result.get()
+
+        # with custom timeout
+        with entrypoint_waiter(container, 'handle_event', timeout=5):
+            ...  # action that dispatches event
+
+        # with callback that waits until entrypoint stops raising
+        def callback(worker_ctx, result, exc_info):
+            if exc_info is None:
+                return True
+
+        with entrypoint_waiter(container, 'handle_event', callback=callback):
+            ...  # action that dispatches event
+
+    """
+    if not get_extension(container, Entrypoint, method_name=method_name):
         raise RuntimeError("{} has no entrypoint `{}`".format(
-            container.service_name, entrypoint))
-    if get_extension(container, EntrypointWaiter, entrypoint=entrypoint):
-        raise RuntimeError("Waiter already registered for {}".format(
-            entrypoint))
-
-    # can't mess with dependencies while container is running
-    wait_for_worker_idle(container)
-    container.dependencies.add(waiter)
-
-    try:
-        yield
-        exc = waiter.Timeout(
-            "Entrypoint {}.{} failed to complete within {} seconds".format(
-                container.service_name, entrypoint, timeout)
-        )
-        with eventlet.Timeout(timeout, exception=exc):
-            waiter.wait()
-    finally:
-        wait_for_worker_idle(container)
-        container.dependencies.remove(waiter)
+            container.service_name, method_name))
+
+    waiter_callback = callback
+    waiter_result = WaitResult()
+
+    def on_worker_result(worker_ctx, result, exc_info):
+        complete = False
```

Comment:
still trying to completely wrap my head around this. so the two new waiters in this pr (`entrypoint_waiter` and `wait_for_call`) have different, incompatible apis? (understandable because of how they came about, and one being more specialised, but seems a shame). could we try to think about ways to change one or both apis a little to make them more similar (or even make e_w a special case of wfc?). that might make things a bit more user-friendly

or am i missing something?